core/src/main/scala/org/apache/spark/storage/memory/MemoryStore.scala

```diff
@@ -384,14 +385,37 @@ private[spark] class MemoryStore(
     }
   }
 
+  private def maybeReleaseResources(entry: MemoryEntry[_]): Unit = {
+    entry match {
+      case SerializedMemoryEntry(buffer, _, _) => buffer.dispose()
+      case DeserializedMemoryEntry(values: Array[Any], _, _) => maybeCloseValues(values)
+      case _ =>
+    }
+  }
+
+  private def maybeCloseValues(values: Array[Any]): Unit = {
+    values.foreach {
+      case closable: AutoCloseable =>
+        safelyCloseValue(closable)
+      case _ =>
+    }
+  }
+
+  private def safelyCloseValue(closable: AutoCloseable): Unit = {
+    try {
+      closable.close()
+    } catch {
+      case ex: Exception => logWarning(s"Failed to close AutoClosable $closable", ex)
+    }
+  }
+
   def remove(blockId: BlockId): Boolean = memoryManager.synchronized {
     val entry = entries.synchronized {
       entries.remove(blockId)
     }
     if (entry != null) {
-      entry match {
-        case SerializedMemoryEntry(buffer, _, _) => buffer.dispose()
-        case _ =>
+      if (blockId.isBroadcast) {
+        maybeReleaseResources(entry)
```

Comment:
In this case, what happen when the blockId is not broadcast? The existing cleaning-up will not be called.

Reply:
@dbtsai thanks - That whole "my day job vs. OSS" rush to fix. Will take care of it correctly and push an update.